staging/src/k8s.io/client-go/tools/reference/ref.go

```diff
@@ -73,27 +55,32 @@ func GetReference(scheme *runtime.Scheme, obj runtime.Object) (*v1.ObjectReferen
 		listMeta = objectMeta
 	}
 
-	// if the object referenced is actually persisted, we can also get version from meta
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	kind := gvk.Kind
 	version := gvk.GroupVersion().String()
-	if len(version) == 0 {
-		selfLink := listMeta.GetSelfLink()
-		if len(selfLink) == 0 {
-			return nil, ErrNoSelfLink
-		}
-		selfLinkUrl, err := url.Parse(selfLink)
+
+	// If object meta doesn't contain data about kind and/or version,
+	// we are falling back to scheme.
+	//
+	// TODO: This doesn't work for CRDs, which are not registered in scheme.
```

Comment:
This PR doesn't really make this worse - I don't think we will ever have a situation that kind was set, but version wasn't. Which means we were falling back for kind to scheme anyway.

Reply:
that means anything calling GetReference no longer works for custom object references, right? the event recorder looks to call this method with arbitrary objects

Reply:
I actually claim that this didn't change - I wasn't able to find any case where kind was set and group/version wasn't (in the object we were trying to reference).
And if the kind was also missing, the to get it, we were already falling back to scheme, which was failing (because CRDs aren't registered there). This is what I have on my mind:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/reference/ref.go#L57
[and the comment line above "this is wrong" I guess is exactly about that case, or am I missing something?]

Reply:
I guess scheme can actually pull a gvk from a custom resource... see https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L236-L248

Reply:
Interesting (doesn't seem like something extremely useful though).